WordPress/gutenberg
# Tabs: improve focus behavior

>[!NOTE]
>We're talking about `Tabs`, which is a new version of `TabPanel`. There are subcomponents whose names involve the word 'Tab' or 'TabPanel'. There are actual elements with `tab` and `tabpanel` roles. To top it all off, this PR focuses on the behavior of the [Tab] key on the keyboard. Basically there are a lot of ways to use the word "tab" here, so let me know if anything fails to make sense 😉

## What?
Add focus styles and a new prop to Tabs.TabPanel to declare whether or not the `tabpanel` itself should be focusable.

## Why?
[The relevant ARIA guidance for the Tabs Pattern](https://www.w3.org/WAI/ARIA/apg/patterns/tabs/#keyboardinteraction) states:

> When the tab list contains the focus, [Tab] moves focus to the next element in the page tab sequence outside the tablist, which is the tabpanel unless the first element containing meaningful content inside the tabpanel is focusable.

## How?
Currently, the [Tab] key moves focus from the `tablist` to the currently displayed `tabpanel`, regardless of the presence of focusable child elements within that `tabpanel`. 

We've decided to maintain the existing behavior as the default to remain consistent with the behavior of other components (`Modal`, for example). Along with that, this PR exposes a `focusable` prop on the `Tabs.TabPanel` subcomponent to better control this behavior.

The new prop will default to `true`, making the `tabpanel` focusable, so it will be the next tabstop when using the [Tab] key to navigate off of the `tablist`. When the new prop is `false`, the `tabpanel` is no longer focusable. Focus will instead jump to the first focusable element in the DOM. That could be a button or input within the `tabpanel`, or if none are present, whatever the next tabstop is elsewhere on the page. We aren't targeting any specific elements to focus, as we don't have a reliable means of determining which elements contain "meaningful content" from one implementation to the next.

While this doesn't perfectly align with the guidance described above, it feels like a good compromise of consistency with other components and flexibility for consumers.

## Testing Instructions
- Launch Storybook and select the Default story
- Move the focus to the main canvas, either by clicking on the empty space of the main pane, or by pressing [Tab] followed by [Return/Enter]
- Use the [Tab] key to focus the first `Tab`. You may need to press it twice to cycle over Storybook's "Return to sidebar" button.
- Once the first `Tab` has focus (it will gain an outline), press the [Tab] key again, and confirm that the focus moves to the `tabpanel` element.
- While the `tabpanel` is focused, inspect the element and confirm the component's new `:focus-visible` style is overriding the browser defaults.
- Select the "TabPanels with focusable={ false }" story. Repeat the above steps to test the tabstop behavior for each tab. Remember to use your arrow keys to navigate between `Tab`s within the `tablist`
- For each `Tab`, confirm that the `tabpanel` does _not_ get focus. The focus should instead cycle through any focusable elements within the TabPanel contents, then to the button below (and outside of) the component.
- For the third tab in the story, confirm that the [Tab] key skips the focus directly down to the button lower on the page due to the absence of any focusable items within the displayed `tabpanel`.
